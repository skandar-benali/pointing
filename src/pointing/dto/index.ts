export * from './check-in-out.dto';
