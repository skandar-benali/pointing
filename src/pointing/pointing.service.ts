import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { Pointing } from './entities';
import { PrismaService } from '../prisma.service';

@Injectable()
export class PointingService {
  constructor(private prisma: PrismaService) {}

  async create(
    employee: number,
    pointing: Prisma.PointingCreateWithoutEmployeeInput,
  ): Promise<Pointing> {
    return this.prisma.pointing.create({
      data: {
        ...pointing,
        employee: {
          connect: { id: employee },
        },
      },
    });
  }

  async findOne(id: number): Promise<Pointing> {
    return this.prisma.pointing.findUnique({
      where: { id },
    });
  }
  async update(
    id: number,
    pointing: Prisma.PointingUpdateWithoutEmployeeInput,
  ): Promise<Pointing> {
    return this.prisma.pointing.update({
      where: { id },
      data: pointing,
    });
  }

  async lastPointing(employee): Promise<Pointing | null> {
    const all = await this.prisma.pointing.findMany({
      where: { idEmployee: employee, checkOut: null },
      orderBy: [{ createdAt: 'desc' }],
    });
    return all[0];
  }
}
