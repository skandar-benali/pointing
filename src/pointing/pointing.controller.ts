import { Controller, Post, Body } from '@nestjs/common';
import { CheckInOutDto } from './dto';
import { PointingCore } from './pointing.core';

@Controller('pointing')
export class PointingController {
  constructor(private readonly pointingCore: PointingCore) {}

  @Post('check-in')
  checkIn(@Body() payload: CheckInOutDto) {
    const { employee, comment } = payload;
    return this.pointingCore.checkIn(employee, comment);
  }

  @Post('check-out')
  checkOut(@Body() payload: CheckInOutDto) {
    const { employee, comment } = payload;
    return this.pointingCore.checkOut(employee, comment);
  }
}
