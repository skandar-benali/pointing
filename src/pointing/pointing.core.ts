import { Injectable } from '@nestjs/common';
import { Pointing } from './entities';
import { PointingService } from './pointing.service';

function timestamp(date: Date): number {
  const d = new Date(date);
  return d.getTime();
}
@Injectable()
export class PointingCore {
  constructor(private pointingService: PointingService) {}
  async checkIn(employee: number, comment?: string): Promise<Pointing> {
    // create new pointing raw
    return await this.pointingService.create(employee, {
      checkIn: new Date(),
      comment,
    });
  }

  async checkOut(employee: number, comment?: string): Promise<Pointing> {
    const checkOutDate = new Date();
    // get the last pointing for this employee
    let pointing = await this.pointingService.lastPointing(employee);
    if (!pointing) pointing = await this.pointingService.create(employee, {});

    // calculate the period of work
    let period = null;
    if (pointing.checkIn) {
      period =
        (timestamp(checkOutDate) - timestamp(pointing.checkIn)) / 3600000; // 1000 * 60 * 60 to have the period in hours
    }

    // update it with date
    return await this.pointingService.update(pointing.id, {
      checkOut: new Date(),
      comment,
      period,
    });
  }
}
