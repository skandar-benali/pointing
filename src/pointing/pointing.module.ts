import { Module } from '@nestjs/common';
import { PointingService } from './pointing.service';
import { PointingController } from './pointing.controller';
import { PrismaService } from '../prisma.service';
import { PointingCore } from './pointing.core';

@Module({
  controllers: [PointingController],
  providers: [PointingService, PrismaService, PointingCore],
})
export class PointingModule {}
