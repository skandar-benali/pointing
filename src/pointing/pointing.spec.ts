import { Test } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { PointingModule } from './pointing.module';
import { PointingService } from './pointing.service';

describe('/pointing/check-in POST checkin', () => {
  let app: INestApplication;
  const checkIn = {
    id: 1,
    checkIn: '2021-12-16T10:53:12.522Z',
    checkOut: null,
    comment: 'test comment',
    createdAt: '2021-12-16T10:53:12.522Z',
    updatedAt: null,
  };
  const pointingService = {
    create: () => checkIn,
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [PointingModule],
    })
      .overrideProvider(PointingService)
      .useValue(pointingService)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  test('should return the created employee', async () => {
    // console.log(res.);

    return request(app.getHttpServer())
      .post('/pointing/check-in')
      .send({
        employee: 1,
        comment: 'test comment',
      })
      .expect(201)
      .then((response) => {
        expect(response.body).toMatchSnapshot();
      });
  });

  afterAll(async () => {
    await app.close();
  });
});
