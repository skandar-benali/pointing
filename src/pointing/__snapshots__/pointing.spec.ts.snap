// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/pointing/check-in POST checkin should return the created employee 1`] = `
Object {
  "checkIn": "2021-12-16T10:53:12.522Z",
  "checkOut": null,
  "comment": "test comment",
  "createdAt": "2021-12-16T10:53:12.522Z",
  "id": 1,
  "updatedAt": null,
}
`;
